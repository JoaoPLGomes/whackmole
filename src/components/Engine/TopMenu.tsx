import React from 'react';
import MoleButton from '../shared/MoleButton';

interface TopMenuProps {
  counter: number;
  isRunning: boolean;
  setIsRunning: (isRunning: boolean) => void;
  playerScore: number;
}

const TopMenu = ({ counter, isRunning, setIsRunning, playerScore }: TopMenuProps): JSX.Element => {
  return (
    <div className="w-full absolute top-0 p-4 flex justify-between">
      <p className="text-4xl text-white">Time left: {counter}</p>
      {isRunning && <MoleButton onClick={() => setIsRunning(!isRunning)}>Pause</MoleButton>}
      <p className="text-4xl text-white">Score : {playerScore}</p>
    </div>
  );
};

export default TopMenu;
