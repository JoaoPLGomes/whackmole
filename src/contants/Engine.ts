export const INITIAL_STATE_ENGINE = {
  '1': false,
  '2': false,
  '3': false,
  '4': false,
  '5': false,
  '6': false,
  '7': false,
  '8': false,
  '9': false,
  '10': false,
  '11': false,
  '12': false,
};

export const INITIAL_TIMER = 120;