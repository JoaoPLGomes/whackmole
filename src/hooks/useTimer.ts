import React, { useEffect } from "react";



const useTimer = (timer: number) => {
  const [isRunning, setIsRunning] = React.useState<boolean>(true);
  const [counter, setCounter] = React.useState<number>(timer * 10);

  useEffect(() => {
    counter > 0 && isRunning && setTimeout(() => setCounter(counter - 1), 100);
  }, [counter, isRunning]);

  return ({
    counter,
    isRunning,
    setIsRunning,
  }
  )
}

export default useTimer;