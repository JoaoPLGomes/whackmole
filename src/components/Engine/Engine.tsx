import { INITIAL_STATE_ENGINE } from '@/contants/Engine';
import useTimer from '@/hooks/useTimer';
import React, { useEffect } from 'react';
import Mole from '../Mole/Mole';

const Engine = (): JSX.Element => {
  const { counter, setIsRunning, isRunning } = useTimer(120);
  const [moles, setMoles] = React.useState<Record<string, boolean>>(INITIAL_STATE_ENGINE);

  const onHammerDown = (key: string) => {
    if (moles[key] && isRunning) {
      setMoles({ ...moles, [key]: false });
    }
  };

  useEffect(() => {
    if (isRunning) {
      const randomKey = Object.keys(moles)[Math.floor(Math.random() * 12)];
      if (!moles[randomKey]) {
        setMoles({ ...moles, [randomKey]: true });
      }
    }
  }, [counter]);

  return (
    <div className="h-2/3 w-2/3 grid grid-cols-4 grid-rows-3 z-0">
      {Object.keys(moles).map((key) => (
        <Mole key={key} isUp={moles[key]} onHammerDown={() => onHammerDown(key)} setDown={() => setMoles({ ...moles, [key]: false })} />
      ))}
      {counter}
      <button onClick={() => setIsRunning(!isRunning)}></button>
    </div>
  );
};

export default Engine;
