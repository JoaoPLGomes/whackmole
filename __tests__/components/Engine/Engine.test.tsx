import Engine from '@/components/Engine/Engine';
import { render, screen, waitFor } from '@testing-library/react';
import * as timer from '@/hooks/useTimer';

jest.mock('@/hooks/useTimer');

jest.spyOn(require('next/router'), 'useRouter').mockImplementation(() => ({
  route: '/',
  pathname: '',
  query: '',
  asPath: '',
  push: jest.fn(),
}));

describe('Engine', () => {
  it('renders correctly', () => {
    const { container } = render(<Engine />);
    expect(container).toMatchSnapshot();
  });
  it('renders correct text', () => {
    jest
      .spyOn(timer, 'useTimer')
      .mockImplementationOnce(() => ({ counter: 10, isRunning: false, setIsRunning: jest.fn(), resetTimer: jest.fn() }));
    const { rerender } = render(<Engine />);
    const text = screen.queryByText('Whack a Mole');
    expect(text).toBeInTheDocument();
    jest
      .spyOn(timer, 'useTimer')
      .mockImplementationOnce(() => ({ counter: 10, isRunning: true, setIsRunning: jest.fn(), resetTimer: jest.fn() }));
    rerender(<Engine />);
    const text2 = screen.queryByText('Whack a Mole');
    expect(text2).not.toBeInTheDocument();
  });
  it('some moles src should change after rerendering', async () => {
    jest
      .spyOn(timer, 'useTimer')
      .mockImplementation(() => ({ counter: 10, isRunning: false, setIsRunning: jest.fn(), resetTimer: jest.fn() }));
    const { rerender } = render(<Engine />);
    const moles = screen.getAllByRole('img');
    expect(moles.every((mole) => mole.getAttribute('src') === '/WAM_Hole.png')).toBeTruthy();
    jest
      .spyOn(timer, 'useTimer')
      .mockImplementation(() => ({ counter: 0, isRunning: true, setIsRunning: jest.fn(), resetTimer: jest.fn() }));
    rerender(<Engine />);
    await waitFor(() => expect(moles.some((mole) => mole.getAttribute('src') === '/WAM_Mole.png')).toBeTruthy());
  });
});
