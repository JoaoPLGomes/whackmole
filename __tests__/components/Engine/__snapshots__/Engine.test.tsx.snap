// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine renders correctly 1`] = `
<div>
  <div
    class="absolute bg-stone-800 text-white w-fit h-fit p-8 z-10 rounded-xl text-center flex flex-col gap-4"
  >
    <h1
      class="text-4xl"
    >
      Whack a Mole
    </h1>
    <button
      aria-label="Continue/Start"
      class="bg-stone-600 p-2 rounded-xl"
    >
      Start
    </button>
    <button
      class="bg-stone-600 p-2 rounded-xl"
    >
      Check Leaderboards
    </button>
  </div>
  <div
    class="w-full absolute top-0 p-4 flex justify-between"
  >
    <p
      class="text-4xl text-white"
    >
      Time left: 
      120
    </p>
    <p
      class="text-4xl text-white"
    >
      Score : 
      0
    </p>
  </div>
  <div
    class="h-2/3 w-2/3 grid grid-cols-4 grid-rows-3 z-0"
  >
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
    <div
      class="flex justify-center items-center"
    >
      <img
        src="/WAM_Hole.png"
      />
    </div>
  </div>
</div>
`;
